src/serde/parse_atom.rs

```diff
@@ -9,40 +9,44 @@ const MAX_SINGLE_BYTE: u8 = 0x7f;
 /// decode the length prefix for an atom. Atoms whose value fit in 7 bits
 /// don't have a length prefix, so those should be handled specially and
 /// never passed to this function.
-pub fn decode_size<R: Read>(f: &mut R, initial_b: u8) -> Result<u64> {
+pub fn decode_size_with_offset<R: Read>(f: &mut R, initial_b: u8) -> Result<(u8, u64)> {
     debug_assert!((initial_b & 0x80) != 0);
     if (initial_b & 0x80) == 0 {
         return Err(internal_error());
     }
 
-    let mut bit_count = 0;
+    let mut atom_start_offset = 0;
     let mut bit_mask: u8 = 0x80;
     let mut b = initial_b;
     while b & bit_mask != 0 {
-        bit_count += 1;
+        atom_start_offset += 1;
         b &= 0xff ^ bit_mask;
         bit_mask >>= 1;
     }
     let mut size_blob: Vec<u8> = Vec::new();
-    size_blob.resize(bit_count, 0);
+    size_blob.resize(atom_start_offset, 0);
     size_blob[0] = b;
-    if bit_count > 1 {
-        let remaining_buffer = &mut size_blob[1..];
+    if atom_start_offset > 1 {
+        let remaining_buffer = &mut size_blob[1..atom_start_offset];
```

Comment:
adding this upper end to the range suggests that it's relevant. that there are situations where the `size_blob` is longer that this. I don't think that's the case, and it's a bit confusing to add it here. But perhaps there's some case I'm missing. (if so, I think it warrants a comment in the code)